neqo-transport/src/send_stream.rs

```diff
@@ -1063,61 +1106,174 @@ impl ::std::fmt::Display for SendStream {
 }
 
 #[derive(Debug, Default)]
-pub(crate) struct SendStreams(IndexMap<StreamId, SendStream>);
+pub(crate) struct SendStreams {
+    map: IndexMap<StreamId, SendStream>,
+
+    // What we really want is a Priority Queue that we can do arbitrary
+    // removes from (so we can reprioritize). BinaryHeap doesn't work,
+    // because there's no remove().  BTreeMap doesn't work, since you can't
+    // duplicate keys.  PriorityQueue does have what we need, except for an
+    // ordered iterator that doesn't consume the queue.  So we roll our own.
+
+    // So we use a sorted Vec<> for the regular streams (that's usually all of
+    // them), and then a BTreeMap of an entry for each SendOrder value, and
+    // for each of those entries a HashSet of the stream_ids at that
+    // sendorder.  In most cases (such as stream-per-frame), there will be
+    // a single stream at a given sendorder.
+
+    // These both store stream_ids, which need to be looked up in 'map'.
+    // This avoids the complexity of trying to hold references to the
+    // Streams which are owned by the IndexMap.
+    sendordered: BTreeMap<SendOrder, HashSet<StreamId>>,
+    regular: Vec<StreamId>, // streams with no SendOrder set, sorted in stream_id order
+}
 
 impl SendStreams {
     pub fn get(&self, id: StreamId) -> Res<&SendStream> {
-        self.0.get(&id).ok_or(Error::InvalidStreamId)
+        self.map.get(&id).ok_or(Error::InvalidStreamId)
     }
 
     pub fn get_mut(&mut self, id: StreamId) -> Res<&mut SendStream> {
-        self.0.get_mut(&id).ok_or(Error::InvalidStreamId)
+        self.map.get_mut(&id).ok_or(Error::InvalidStreamId)
     }
 
     pub fn exists(&self, id: StreamId) -> bool {
-        self.0.contains_key(&id)
+        self.map.contains_key(&id)
     }
 
     pub fn insert(&mut self, id: StreamId, stream: SendStream) {
-        self.0.insert(id, stream);
+        self.map.insert(id, stream);
+	// This normally is only called when a new stream is created.  If
+	// so, because of how we allocate StreamIds, it should always have
+	// the largest value.  This means we can just append it to the
+	// regular vector.  However, if we were ever to change this
+	// invariant, things would break subtly.
+
+	// We could add assertions, or we can try to insert at the end and
+	// if not fall back to binary-search insertion
+	if let Some(last) = self.regular.last() {
+	    if id > *last {
+		self.regular.push(id);
+		return;
+	    }
+	}
+	self.insert_regular(&id);
```

Comment:
```suggestion
	match self.regular.last() {
	  Some(last) if id > *last => self.regular.push(id);
	  _ => self.insert_regular(&id);
	}
```
There you go haxx.  You can also do

```suggestion
	if matches!(self.regular.last(), Some(last) if id > *last)) {
	  self.regular.push(id);
	} else {
	  self.insert_regular(&id);
	}
```